### PR TITLE
[DONOT MERGE]Cache thirdparty pacakge on CI system, TEST CI

### LINKED ISSUE #9046: Turn on the cmake flag WITH_DISTRIBUTE on CI system
it's better to cache the third-party package on CI so that we can improve CI performance and turn on `WITH_DISTRIBUTE` cmake flag as also.

We can't turn on `WITH_DISTRIBUTE` currently is just because building `gRPC` dependency would task very very long time.

### PR DESCRIPTION
**Pull request by Yancey1989**
Fixed #9046 

Test task on CI: https://paddleci.ngrok.io/viewType.html?buildTypeId=Paddle_DistributedUnitTestNightly&branch_Paddle=9150&tab=buildTypeStatusDiv

### CODE REVIEW

**Comment by helinwang** [comment]
Thanks for the PR! TeamCity CI is very special in the sense that `nvidia-docker` command is run **outside** of docker container. Perhaps we should not let TeamCity CI run arbitrary script outside of docker, otherwise it's a big security risk.

For example, currently `build.sh` is run inside of docker, but the command to start the container (`nvidia-docker run ...`) is configured by admin in TeamCity.

Maybe this PR can be documented in the wiki, and configured in the TeamCity management settings?

File: paddle/scripts/teamcity/pr_build.sh [hunk 1]
```diff
@@ -0,0 +1,116 @@
...
+    -e "WITH_GPU=%WITH_GPU%"\
+    -e "CUDA_ARCH_NAME=Auto"\
+    -e "WITH_AVX=%WITH_AVX%"\
+    -e "WITH_GOLANG=%WITH_GOLANG%"\
+    -e "WITH_TESTING=ON"\
+    -e "WITH_C_API=%WITH_C_API%" \
+    -e "WITH_COVERAGE=ON" \
+    -e "COVERALLS_UPLOAD=ON" \
+    -e "GIT_PR_ID=${PR_ID}" \
+    -e "JSON_REPO_TOKEN=JSUOs6TF6fD2i30OJ5o2S55V8XWv6euen" \
```
Comment (helinwang) [inline paddle/scripts/teamcity/pr_build.sh]: Is this token sensitive data :) ?

**Comment by Yancey1989** [comment]
Thanks for @helinwang! It's a very important reminder. And maybe we can review this PR and then close && move the script to a wiki page after approved.

File: paddle/scripts/teamcity/pr_build.sh [hunk 2]
```diff
@@ -0,0 +1,116 @@
...
+    mkdir -p ${PWD}/build
+    tar zxvf ${tp_cache_file}.tar.gz -C $PWD
+else
+    # clear the older tar files if MD5 has chanaged.
+    rm -rf $tp_cache_dir/*.tar.gz
+fi
+
+# Do not build dev image for now
+
+if [[ "%WITH_UBUNTU_MIRROR%" == "ON" ]]; then
```
Comment (typhoonzero) [inline paddle/scripts/teamcity/pr_build.sh]: Can test if `Dockerfile` is not changed, then we can skip building the image.

File: paddle/scripts/teamcity/pr_build.sh (hunk 1 above)
Comment (Yancey1989) [reply paddle/scripts/teamcity/pr_build.sh]: Sure...Done.

File: paddle/scripts/teamcity/pr_build.sh (hunk 2 above)
Comment (Yancey1989) [reply paddle/scripts/teamcity/pr_build.sh]: May not, a PR job may run on any node of the CI cluster, for one PR changes the `Dockerfile` and merged, the next PR run another node without changing of `Dockerfile`, and it would use the older dev Docker image, we do not want this. For another side,`docker build` would use the cached image Layer if `Dockerfile` has no changs, would not cost too much time.

File: paddle/scripts/teamcity/pr_build.sh [hunk 3]
```diff
@@ -0,0 +1,115 @@
...
+apt_mirror='s#http://archive.ubuntu.com/ubuntu#mirror://mirrors.ubuntu.com/mirrors.txt#g'
+
+export PADDLE_DEV_NAME=paddlepaddle/paddle:dev
+
+nvidia-docker run -i --rm -v $PWD:/paddle ${PADDLE_DEV_NAME} rm -rf /paddle/build
+
+# Use the cached thirdparty package if the MD5 is mathes.
+md5_content=$(cat \
+              ${PWD}/cmake/external/*.cmake \
+              ${PWD}/paddle/scripts/teamcity/pr_build.sh \
```
Comment (helinwang) [inline paddle/scripts/teamcity/pr_build.sh]: Maybe don't need this line anymore it we don't check-in `pre-build.sh` to our repo.

File: paddle/scripts/teamcity/pr_build.sh [hunk 4]
```diff
@@ -0,0 +1,115 @@
...
+
+apt_mirror='s#http://archive.ubuntu.com/ubuntu#mirror://mirrors.ubuntu.com/mirrors.txt#g'
+
+export PADDLE_DEV_NAME=paddlepaddle/paddle:dev
+
+nvidia-docker run -i --rm -v $PWD:/paddle ${PADDLE_DEV_NAME} rm -rf /paddle/build
+
+# Use the cached thirdparty package if the MD5 is mathes.
+md5_content=$(cat \
+              ${PWD}/cmake/external/*.cmake \
```
Comment (helinwang) [inline paddle/scripts/teamcity/pr_build.sh]: Smart!

File: paddle/scripts/teamcity/pr_build.sh (hunk 2 above)
Comment (helinwang) [reply paddle/scripts/teamcity/pr_build.sh]: Does it make sense to tag the docker image with the hash of the Dockerfile? (and remove all other tags when a PR builds)
E.g.,
docker image:
paddlepaddle/paddle:dev-000000
paddlepaddle/paddle:dev-000001

If the hash is 000001, then use `paddlepaddle/paddle:dev-000001` and remove `paddlepaddle/paddle:dev-000000` to free up space.

File: paddle/scripts/teamcity/pr_build.sh (hunk 3 above)
Comment (Yancey1989) [reply paddle/scripts/teamcity/pr_build.sh]: Done.

File: paddle/fluid/operators/CMakeLists.txt [hunk 5]
```diff
@@ -156,7 +156,8 @@ if(WITH_DISTRIBUTE)
     set_source_files_properties(recv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
     op_library(listen_and_serv_op DEPS ${DISTRIBUTE_DEPS})
     set_source_files_properties(listen_and_serv_op.cc PROPERTIES COMPILE_FLAGS ${DISTRIBUTE_COMPILE_FLAGS})
-    cc_test(test_send_recv SRCS send_recv_op_test.cc DEPS send_op listen_and_serv_op sum_op executor)
+    # FIXME(Yancey1989): fix unit test
+    # cc_test(test_send_recv SRCS send_recv_op_test.cc DEPS send_op listen_and_serv_op sum_op executor)
```
Comment (helinwang) [inline paddle/fluid/operators/CMakeLists.txt]: It's better to separate this to another PR next time :)

File: python/paddle/fluid/tests/unittests/test_recv_op.py [hunk 6]
```diff
@@ -21,6 +21,8 @@
 import os, sys
 import time
 
+sys.exit(0)
```
Comment (helinwang) [inline python/paddle/fluid/tests/unittests/test_recv_op.py]: Maybe add a TODO here, otherwise it's possible to forget.

File: paddle/scripts/teamcity/pr_build.sh [hunk 7]
```diff
@@ -0,0 +1,125 @@
...
+# Check the MD5 of the Dockerfile of dev Docker image,
+# if there is no current Docker iamges, delete all the older 
+# dev Docker images to free up disk space.
+
+md5_dev_dockerfile=$(cat ${PWD}/Dockerfile | md5sum | awk '{print $1}')
+export PADDLE_DEV_NAME=paddlepaddle/paddle-dev:${md5_dev_dockerfile}
+
+dev_image_id=$(docker images paddlepaddle/paddle-dev:${md5_dev_dockerfile} -q)
+if [[ ${dev_image_id} == "" ]]; then
+    docker rmi $(docker images paddlepaddle/paddle-dev -q) || true
```
Comment (helinwang) [inline paddle/scripts/teamcity/pr_build.sh]: Smart of using `paddlepaddle/paddle-dev`!

File: paddle/fluid/operators/CMakeLists.txt (hunk 5 above)
Comment (Yancey1989) [reply paddle/fluid/operators/CMakeLists.txt]: Okay, and I add some comments at https://github.com/PaddlePaddle/Paddle/issues/9046#issuecomment-375162347, there are still some bugs while building with `WITH_DISTRIBUTE=ON.`

**Comment by Yancey1989** [comment]
Thanks for @helinwang 's reviewing, I moved the build script to the wiki page: https://github.com/PaddlePaddle/Paddle/wiki/PaddlePaddle-CI-Build-Scripts.

**Comment by helinwang** [comment]
Thank you!!!
